onnxruntime/core/providers/cuda/reduction/reduction_functions.cu

```diff
@@ -1,87 +1,145 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#include "core/providers/cuda/reduction/reduction_functions.h"
+
 #include <algorithm>
+
 #include <cuda.h>
 #include <cuda_fp16.h>
-#include "core/providers/cuda/cu_inc/common.cuh"
-#include "core/providers/cuda/atomic/common.cuh"
-#include "reduction_functions.h"
-#include "reduction_utils.cuh"
 
-#define NUM_ELEMENTS_PER_THREAD 4
-#define NUM_WARPS_PER_BLOCK 8
-#define MAX_NUM_BLOCKS 256
-
-#define ALL_ONE_MASK 0xFFFFFFFF
-#define ONE_MASK 0x00000001
+#include "core/common/common.h"
+#include "core/providers/cuda/atomic/common.cuh"
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/cuda/shared_inc/cuda_utils.h"
+#include "core/providers/cuda/reduction/reduction_utils.cuh"
 
 namespace onnxruntime {
 namespace cuda {
 
-std::pair<int, int> compute_block_size(int size) {
+namespace detail {
+constexpr auto MAX_NUM_ELEMENTS_PER_THREAD = 4;
+constexpr auto MAX_NUM_WARPS_PER_BLOCK = 8;
+constexpr auto MAX_NUM_BLOCKS_IN_GRID_ROW = 256;
+constexpr auto MAX_NUM_GRID_ROWS = 32768;
+
+dim3 compute_block_dim(int row_size) {
   int x = GPU_WARP_SIZE;
-  int y = std::min(NUM_WARPS_PER_BLOCK, std::max(1, size / (NUM_ELEMENTS_PER_THREAD * GPU_WARP_SIZE)));
-  return std::make_pair(x, y);
+  int y = std::min(MAX_NUM_WARPS_PER_BLOCK, std::max(1, row_size / (MAX_NUM_ELEMENTS_PER_THREAD * x)));
+  return dim3(x, y);
 }
 
-int compute_grid_size(int size) {
-  const auto block = compute_block_size(size);
-  return std::min(MAX_NUM_BLOCKS, std::max(1, size / (NUM_ELEMENTS_PER_THREAD * block.first * block.second)));
+std::pair<dim3, dim3> compute_grid_and_block_dims(int num_rows, int row_size) {
+  const auto block_dim = compute_block_dim(row_size);
+  const auto grid_x =
+      std::min<int>(
+          MAX_NUM_BLOCKS_IN_GRID_ROW,
+          std::max<int>(1, row_size / (MAX_NUM_ELEMENTS_PER_THREAD * block_dim.x * block_dim.y)));
+  const auto grid_y = std::min(MAX_NUM_GRID_ROWS, num_rows);
+  const dim3 grid_dim(grid_x, grid_y);
+  return {grid_dim, block_dim};
+}
+
+uintptr_t round_up_to_aligned(uintptr_t original, size_t alignment) {
+  assert((alignment & (alignment - 1)) == 0);
+  const size_t alignment_mask = ~(alignment - 1);
+  return (original + alignment - 1) & alignment_mask;
+}
+
+size_t compute_reduce_all_in_rows_buffer_size(int element_size, int element_alignment, int num_rows, int row_size) {
+  const auto grid_dim = compute_grid_and_block_dims(num_rows, row_size).first;
+
+  size_t buffer_size{};
+
+  // at the beginning, for sizing purposes, assume we are aligned
+  buffer_size += num_rows * grid_dim.x * element_size;
+
+  buffer_size = round_up_to_aligned(buffer_size, alignof(int));
+  buffer_size += num_rows * sizeof(int);
```

Comment:
I prefer to split this into two buffers, one for counter and the other one for block-level reduction result.

Reply:
Added documentation about intermediate buffer layout as discussed offline.

---
In reply to: [517590891](https://github.com/microsoft/onnxruntime/pull/5639#discussion_r517590891) [](ancestors = 517590891)